shared/src/http_transport.rs

```diff
@@ -0,0 +1,186 @@
+use ethcontract::jsonrpc as jsonrpc_core;
+use futures::{future::BoxFuture, FutureExt};
+use jsonrpc_core::types::{Call, Output, Request, Value};
+use reqwest::{Client, Url};
+use serde::de::DeserializeOwned;
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+};
+use web3::{error::Error as Web3Error, helpers, BatchTransport, RequestId, Transport};
+
+#[derive(Clone, Debug)]
+pub struct HttpTransport {
+    client: Client,
+    inner: Arc<Inner>,
+}
+
+#[derive(Debug)]
+struct Inner {
+    url: Url,
+    id: AtomicUsize,
+}
+
+impl HttpTransport {
+    pub fn new(url: Url) -> Self {
+        Self::with_client(Client::new(), url)
+    }
+
+    pub fn with_client(client: Client, url: Url) -> Self {
+        Self {
+            client,
+            inner: Arc::new(Inner {
+                url,
+                id: AtomicUsize::new(0),
+            }),
+        }
+    }
+
+    pub fn next_id(&self) -> RequestId {
+        self.inner.id.fetch_add(1, Ordering::SeqCst)
+    }
+
+    pub fn new_request(&self) -> (Client, Url) {
+        (self.client.clone(), self.inner.url.clone())
+    }
+}
+
+// Id is only used for logging.
+async fn execute_rpc<T: DeserializeOwned>(
+    client: Client,
+    url: Url,
+    id: RequestId,
+    request: &Request,
+) -> Result<T, Web3Error> {
+    tracing::debug!(
+        "[id:{}] sending request: {:?}",
+        id,
+        serde_json::to_string(&request)?
+    );
+    let response = client.post(url).json(request).send().await.map_err(|err| {
+        let message = format!("failed to send request: {}", err);
+        tracing::debug!("[id:{}] {}", id, message);
+        Web3Error::Transport(message)
+    })?;
+    let status = response.status();
+    let text = response.text().await.map_err(|err| {
+        let message = format!("failed to get response body: {}", err);
+        tracing::debug!("[id:{}] {}", id, message);
+        Web3Error::Transport(message)
+    })?;
+    // Log the raw text before decoding to get more information on responses that aren't valid
+    // json. Debug encoding so we don't get control characters like newlines in the output.
+    tracing::debug!("[id:{}] received response: {:?}", id, text.trim());
```

Comment:
I'd like to discuss this design choices. Debug encoding makes it harder to read the response directly in the logs because everything is one long line. On the other hand I do think it is necessary for the raw text body as the comment mentions.
An alternative could be to decode into serde_json::Value first and pretty print that. This has the advantage the output is struct-like which makes it easier to read. Only if we can't decode into Value first we would debug log the raw text.
What do you think?

Reply:
Personally, I am for this change. It also would resolve the issue @bh2smith opened here: https://github.com/gnosis/gp-v2-services/issues/669.

Reply:
If I read the issue right, you are in favor of keeping the debug log so the output stays in one line. (correct me if i am wrong)